strapi/strapi
# Type System v1

### What does it do?

#### Initialize a new type system for Strapi

Initialize modules and namespaces to export core types and utilities from Strapi.
Core types include the basic building blocks used to create powerful APIs:
- [Namespaces](https://github.com/strapi/strapi/blob/features/type-system/packages/core/strapi/lib/types/core/namespace.d.ts)
- [UIDs](https://github.com/strapi/strapi/blob/features/type-system/packages/core/strapi/lib/types/core/uid.d.ts)
- Dynamic Registries
- Models ([schema](https://github.com/strapi/strapi/blob/features/type-system/packages/core/strapi/lib/types/core/schemas) [component, content-type], [attributes](https://github.com/strapi/strapi/blob/features/type-system/packages/core/strapi/lib/types/core/attributes))

#### Updated the type generator to match the type system

- Now generate one file per registry (components, content types, etc...)
- _todo: generate types for services, controllers, policies, and middlewares._
- _todo: generate index.d.ts files to load automatically registries augmentations_ 

#### Updated the data transfer package to use the created types

### Why is it needed?

In order to provide a better TS developer experience for Strapi users, we need to rely on solid foundations.
The goal of this type system is to act both as a toolbelt and a set of conventions for TypeScript adoption throughout the codebase.

While this is only the first version of this type system and while it should be treated as an experimental release (things might change regarding type definitions in the coming months), it should set a clear objective related to our ambitions with TypeScript.

### How to test it?

- `cd examples/getstarted`
- `yarn run strapi ts:generate-types`
- Two files have been generated at the root of getstarted: `types/shared/registries/component.d.ts` and `types/shared/registries/contentTypes.d.ts` and should contain type definition and dynamic registries augmentation
- You can then use the different namespaces exported from strapi/strapi to build custom definition based on them.
eg:
```ts
import type { Attribute, Common, Utils } from '@strapi/strapi';

declare function find<T extends Common.UID.ContentType, F extends Attribute.GetKeys<T>[]>(
  uid: T,
  fields: F
): Attribute.GetValues<T, Utils.Array.Values<F>>;

find('api::country.country', ['code']).code;
find('api::restaurant.restaurant', ['code']).code;
find('api::category.category', ['date', 'name']).date;

```

### Related issue(s)/PR(s)

Let us know if this is related to any issue/pull request
